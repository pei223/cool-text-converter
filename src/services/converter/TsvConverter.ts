import { ConvertibleMarkdownFormats } from "../consts/format";
import { MarkdownUnrecognizedError } from "../errors/types";
import { writeMarkdownTable } from "./tableWriter";
import { MarkdownConverter } from "./types";

export class TsvConverter implements MarkdownConverter {
  recognize(v: string): void {
    // 1行目にタブがあるかで判断.
    // かなり誤検知がありそうなので優先度は低めにしておく.
    const firstLine = v.split("\n")[0];
    if (!firstLine.includes("\t")) {
      throw new MarkdownUnrecognizedError(v, "TSV", `Not contained tab`);
    }
  }
  getDetailedConverter(): MarkdownConverter | null {
    return null;
  }
  convertToMarkdown(v: string): string {
    const lines = v.split("\n");
    return writeMarkdownTable(
      lines[0].split("\t"),
      lines.length === 1 ? [] : lines.map((v) => v.split("\t")).slice(1)
    );
  }
  formatName(): ConvertibleMarkdownFormats {
    return "TSV";
  }
  formatLabel(): string {
    return "TSV";
  }
}
