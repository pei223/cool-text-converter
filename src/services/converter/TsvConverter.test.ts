import { describe, it, expect } from "vitest";
import { TsvConverter } from "./TsvConverter";

describe("TsvConverter", () => {
  const converter = new TsvConverter();

  describe("recognize", () => {
    it.each([
      "test\ttest",
      "tett\ttest\ttest",
      "head\thead\thead\ntett\ttest\ttest",
    ])("Success(%s)", (v: string) => {
      expect(() => converter.recognize(v)).not.toThrow();
    });
    it.each(["http//hoge", "1234", "invalid"])("Not apply(%s)", (v: string) => {
      expect(() => converter.recognize(v)).toThrow();
    });
  });
  // 中身いちいち正解かくのしんどいのでスナップショットテストにする
  it.each([
    ["test\ttest"],
    ["tett\ttest\ttest"],
    ["head\thead\thead\ntett\ttest\ttest"],
  ])("converteToMarkdown(%s) - (%s)", (v: string) => {
    expect(converter.convertToMarkdown(v)).toMatchSnapshot();
  });
});
