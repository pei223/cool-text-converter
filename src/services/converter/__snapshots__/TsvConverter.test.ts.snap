// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TsvConverter > converteToMarkdown(head	head	head
tett	test	test) - (%s) 1`] = `
"| head | head | head |
| ---- | ---- | ---- |
| tett | test | test |"
`;

exports[`TsvConverter > converteToMarkdown(test	test) - (%s) 1`] = `
"| test | test |
| ---- | ---- |
"
`;

exports[`TsvConverter > converteToMarkdown(tett	test	test) - (%s) 1`] = `
"| tett | test | test |
| ---- | ---- | ---- |
"
`;
